// @magicyan/core/dist/functions/timers.cjs
'use strict';

const promises = require('node:timers/promises');

const sleep = Object.assign(promises.setTimeout, {
  seconds: (time) => promises.setTimeout(time * 1e3),
  minutes: (time) => promises.setTimeout(time * 1e3 * 60)
});
function createInterval(options) {
  const { time, run, immediately } = options;
  if (immediately)
    run(() => {
    });
  const timer = setInterval(() => {
    run(() => clearInterval(timer));
  }, time);
  return { timer, stop: () => clearInterval(timer) };
}
function createTimeout(options) {
  const { delay, run } = options;
  const timer = setTimeout(() => run(), delay);
  return { timer, stop: () => clearTimeout(timer) };
}
function createLoopInterval(options) {
  const { run, time, array, immediately } = options;
  let lap = 0;
  let loop = 0;
  return createInterval({
    immediately,
    time,
    run(stop) {
      if (loop === array.length) {
        lap++;
        loop = 0;
      }
      run(array[loop], stop, lap, array);
      loop++;
    }
  });
}

exports.createInterval = createInterval;
exports.createLoopInterval = createLoopInterval;
exports.createTimeout = createTimeout;
exports.sleep = sleep;
