// @magicyan/core/dist/index.cjs
'use strict';

const convert = require('./functions/convert.cjs');
const format = require('./functions/format.cjs');
const math = require('./functions/math.cjs');
const timers = require('./functions/timers.cjs');
const check = require('./functions/check.cjs');
const validation = require('./functions/validation.cjs');



exports.hexToRgb = convert.hexToRgb;
exports.rgbToHex = convert.rgbToHex;
exports.brBuilder = format.brBuilder;
exports.captalize = format.captalize;
exports.limitText = format.limitText;
exports.notFound = format.notFound;
exports.replaceText = format.replaceText;
exports.spaceBuilder = format.spaceBuilder;
exports.parseFloatOrDefault = math.parseFloatOrDefault;
exports.parseIntOrDefault = math.parseIntOrDefault;
exports.random = math.random;
exports.randomNumber = math.randomNumber;
exports.createInterval = timers.createInterval;
exports.createLoopInterval = timers.createLoopInterval;
exports.createTimeout = timers.createTimeout;
exports.sleep = timers.sleep;
exports.equalsIgnoreCase = check.equalsIgnoreCase;
exports.includesIgnoreCase = check.includesIgnoreCase;
exports.isEmail = validation.isEmail;
exports.isNumeric = validation.isNumeric;
exports.isUrl = validation.isUrl;
